tests/federation_room_join_partial_state_test.go

```diff
@@ -513,6 +558,117 @@ func TestPartialStateJoin(t *testing.T) {
+	 * https://github.com/matrix-org/synapse/issues/13146
+	alice.MustSyncUntil(t,
+		client.SyncReq{
+			Filter: buildLazyLoadingSyncFilter(nil),
+		},
+		client.SyncTimelineHasEventID(psjResult.ServerRoom.RoomID, event.EventID()),
+	)
+	*/
+
+	// still, Alice should be able to see the event with an /event request. We might have to try it a few times.
+	start := time.Now()
+	for {
+		if time.Since(start) > time.Second {
+			t.Fatalf("timeout waiting for received event to be visible")
+		}
+		res := alice.DoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", psjResult.ServerRoom.RoomID, "event", event.EventID()})
+		eventResBody := client.ParseJSON(t, res)
+		if res.StatusCode == 200 {
+			t.Logf("Successfully fetched received event %s", event.EventID())
+			break
+		}
+		if res.StatusCode == 404 && gjson.GetBytes(eventResBody, "errcode").String() == "M_NOT_FOUND" {
+			t.Logf("Fetching received event failed with M_NOT_FOUND; will retry")
+			time.Sleep(100 * time.Millisecond)
+			continue
+		}
+		t.Fatalf("GET /event failed with %d: %s", res.StatusCode, string(eventResBody))
+	}
+
+	// fire off a /state_ids request for the last event.
+	// it must either:
+	//   * block because the homeserver does not have full state at the last event
+	//   * or 403 because the homeserver does not have full state yet and does not consider the
+	//     Complement homeserver to be in the room
+	stateIdsResponseChan := make(chan *gomatrixserverlib.RespStateIDs)
+	go func() {
+		stateReq := gomatrixserverlib.NewFederationRequest("GET", "hs1",
+			fmt.Sprintf("/_matrix/federation/v1/state_ids/%s?event_id=%s",
+				url.PathEscape(psjResult.ServerRoom.RoomID),
+				url.QueryEscape(event.EventID()),
+			),
+		)
+		var respStateIDs gomatrixserverlib.RespStateIDs
+		if err := psjResult.Server.SendFederationRequest(deployment, stateReq, &respStateIDs); err != nil {
+			httpErr, ok := err.(gomatrix.HTTPError)
+			t.Logf("%v", httpErr)
+			if ok && httpErr.Code == 403 {
+				stateIdsResponseChan <- nil
+				return
+			}
+			t.Errorf("/state_ids request returned non-200: %s", err)
+			return
+		}
+		stateIdsResponseChan <- &respStateIDs
+	}()
+
+	select {
+	case <-time.After(1 * time.Second):
+		t.Logf("/state_ids request for event %s blocked as expected", event.EventID())
+		break
+	case respStateIDs := <-stateIdsResponseChan:
+		if respStateIDs == nil {
+			t.Logf("/state_ids request for event %s returned 403 as expected", event.EventID())
+		} else {
+			// since we have not yet given the homeserver the full state at the join event and allowed
+			// the partial join to complete, it can't possibly know the full state at the last event.
+			// While it may be possible for the response to be correct by some accident of state res,
+			// the homeserver is still wrong in spirit.
+			t.Fatalf("/state_ids request for event %s did not block when it should have", event.EventID())
```

Comment:
This part tests https://github.com/matrix-org/synapse/issues/13002 . If Synapse does not correctly flag that it does not have partial state in the `CanReceiveEventsWithMissingPrevWithHalfMissingPrevsDuringPartialStateJoin` test case, it will erroneously return state (which was derived via state res and only happens to be correct) instead of blocking.